hassio/cluster/util.py

```diff
@@ -0,0 +1,97 @@
+"""Init file for HassIO util for cluster."""
+
+import json
+import logging
+
+from aiohttp import web
+from jwcrypto import jwk, jwe
+from jwcrypto.common import json_encode
+
+from ..const import JSON_RESULT, RESULT_OK
+
+_LOGGER = logging.getLogger(__name__)
+
+KEY_TYPE = "A128KW"
+ENC_ALG = "A128CBC-HS256"
+
+
+def generate_cluster_key():
+    """Generating random key (master/node)."""
+    key = jwk.JWK.generate(kty="oct", alg=KEY_TYPE)
+    export = json.loads(key.export_symmetric())
+    return export["k"]
+
+
+def get_node_slug(node_name):
+    """Fixing node name."""
+    return node_name.replace(" ", "_")
+
+
+def get_key(key):
+    """Wrapping encryption key."""
+    return jwk.JWK(kty="oct", alg=KEY_TYPE, k=key)
+
+
+def cluster_decrypt(data, plain_key):
+    """Decrypting data."""
+    key = get_key(plain_key)
+    token = jwe.JWE()
+    token.deserialize(data, key=key)
+    return token.payload.decode("utf-8")
+
+
+async def cluster_decrypt_body(request, plain_key):
+    """Decrypting request/response body."""
+    enc_body = await request.text()
+    return cluster_decrypt(enc_body, plain_key)
+
+
+async def cluster_decrypt_body_json(request, plain_key):
+    """Decrypting request/response body into JSON object."""
+    body = await cluster_decrypt_body(request, plain_key)
+    return json.loads(body)
+
+
+async def cluster_decrypt_schema(schema, request, plain_key):
+    """Decrypting request/response body with schema validation."""
+    return schema(await cluster_decrypt_body_json(request, plain_key))
+
+
+def cluster_public_api_process(method):
+    """Wrapping every possible error."""
+
+    async def wrap_api(api, *args, **kwargs):
+        """Return api information."""
+        try:
+            result = await method(api, *args, **kwargs)
+            return web.Response(body=result)
+        # pylint: disable=broad-except
+        except Exception as err:
```

Comment:
This wraps various possible errors, including jwcrypt ones, not from hassio only